commit e4807621b8f41fc4f9fa69f423f1fbf7bba05218
Author: Maximilian Michels <mxm@apache.org>
Date:   Wed Oct 26 16:37:56 2016 +0100

    [FLINK-3813][yarn] wait for CLI to complete before checking output
    
    This closes #2749.

diff --git a/flink-yarn-tests/src/test/java/org/apache/flink/yarn/YARNSessionFIFOITCase.java b/flink-yarn-tests/src/test/java/org/apache/flink/yarn/YARNSessionFIFOITCase.java
index c5659eb..68cc73d 100644
--- a/flink-yarn-tests/src/test/java/org/apache/flink/yarn/YARNSessionFIFOITCase.java
+++ b/flink-yarn-tests/src/test/java/org/apache/flink/yarn/YARNSessionFIFOITCase.java
@@ -81,10 +81,11 @@ public class YARNSessionFIFOITCase extends YarnTestBase {
 	 * Test regular operation, including command line parameter parsing.
 	 */
 	@Test(timeout=60000) // timeout after a minute.
-	public void testDetachedMode() {
+	public void testDetachedMode() throws InterruptedException {
 		LOG.info("Starting testDetachedMode()");
 		addTestAppender(FlinkYarnSessionCli.class, Level.INFO);
-		startWithArgs(new String[]{"-j", flinkUberjar.getAbsolutePath(),
+		Runner runner =
+			startWithArgs(new String[]{"-j", flinkUberjar.getAbsolutePath(),
 						"-t", flinkLibFolder.getAbsolutePath(),
 						"-n", "1",
 						"-jm", "768",
@@ -93,6 +94,8 @@ public class YARNSessionFIFOITCase extends YarnTestBase {
 						"--detached"},
 				"Flink JobManager is now running on", RunTypes.YARN_SESSION);
 
+		// before checking any strings outputted by the CLI, first give it time to return
+		runner.join();
 		checkForLogString("The Flink YARN client has been started in detached mode");
 
 		LOG.info("Waiting until two containers are running");
